arangod/Replication2/ReplicatedLog/Components/StorageManager.cpp

```diff
@@ -0,0 +1,358 @@
+    GuardType&& guard, InMemoryLog result, TermIndexMapping mapResult,
+    std::unique_ptr<StorageOperation> operation)
+    -> futures::Future<arangodb::Result> {
+  guard->spearheadLog = result;
+  guard->spearheadMapping = mapResult;
+  auto f = guard->queue
+               .emplace_back(std::move(operation), std::move(result),
+                             std::move(mapResult))
+               .promise.getFuture();
+  triggerQueueWorker(std::move(guard));
+  return f;
+}
+
+template<typename F>
+auto StorageManager::scheduleOperationLambda(GuardType&& guard,
+                                             InMemoryLog result,
+                                             TermIndexMapping mapResult, F&& fn)
+    -> futures::Future<Result> {
+  struct LambdaOperation : F, StorageOperation {
+    explicit LambdaOperation(F&& fn) : F(std::move(fn)) {}
+    static_assert(std::is_nothrow_invocable_r_v<futures::Future<Result>, F,
+                                                IStorageEngineMethods&>);
+    futures::Future<Result> run(
+        IStorageEngineMethods& methods) noexcept override {
+      return F::operator()(methods);
+    }
+  };
+
+  return scheduleOperation(
+      std::move(guard), std::move(result), std::move(mapResult),
+      std::make_unique<LambdaOperation>(std::forward<F>(fn)));
+}
+
+void StorageManager::triggerQueueWorker(GuardType guard) noexcept {
+  auto const worker = [](GuardType guard,
+                         std::shared_ptr<StorageManager> self) noexcept
+      -> futures::Future<futures::Unit> {
+    LOG_CTX("6efe9", TRACE, self->loggerContext)
+        << "starting new storage worker";
+    while (true) {
+      if (guard->queue.empty()) {
+        LOG_CTX("882a0", TRACE, self->loggerContext)
+            << "stopping storage worker";
+        guard->workerActive = false;
+        break;
+      }
+
+      auto req = std::move(guard->queue.front());
+      guard->queue.pop_front();
+      if (not guard->methods) {
+        guard.unlock();
+        LOG_CTX("4f5e3", DEBUG, self->loggerContext)
+            << "aborting storage operation "
+            << " because log core gone";
+        req.promise.setValue(
+            TRI_ERROR_REPLICATION_REPLICATED_LOG_PARTICIPANT_GONE);
+        guard = self->guardedData.getLockedGuard();
+        continue;
+      }
+      LOG_CTX("e0a6d", TRACE, self->loggerContext)
+          << "executing storage operation";
+      auto f = req.operation->run(*guard->methods);
+      guard.unlock();
+      Result result = co_await asResult(std::move(f));
+      // lock guard to make effects visible
+      if (result.ok()) {
+        LOG_CTX("b6cbf", TRACE, self->loggerContext)
+            << "storage operation completed";
+        // resolve and continue
+        req.promise.setValue(std::move(result));
+        guard = self->guardedData.getLockedGuard();
+        guard->onDiskLog = std::move(req.logResult);
+        guard->onDiskMapping = std::move(req.mappingResult);
```

Comment:
Is it OK to resolve the promise before the value is visible in the data structures? E.g., if we do an appendEntries, followed by an updateCommitIndex; I assume it's OK because the FollowerCommitManager calculates a resolveIndex from both commit index and the visible log?

Reply:
Yes, I think this i fine, because who every looks into the log will see spearhead and not what is on disk.

Reply:
Ah not true. There is `getCommitedLog` 🤔 In that case we have to lock the mutex, update the onDisk state and then unlock, resolve, lock.

Reply:
Fixed it.